cmd/noms/splore/noms_splore.go

```diff
@@ -0,0 +1,307 @@
+	"reflect"
+
+	"github.com/attic-labs/noms/cmd/util"
+	"github.com/attic-labs/noms/go/config"
+	"github.com/attic-labs/noms/go/d"
+	"github.com/attic-labs/noms/go/spec"
+	"github.com/attic-labs/noms/go/types"
+	"github.com/attic-labs/noms/go/util/verbose"
+	humanize "github.com/dustin/go-humanize"
+	flag "github.com/juju/gnuflag"
+	"github.com/skratchdot/open-golang/open"
+)
+
+var (
+	NomsSplore = &util.Command{
+		Run:       run,
+		Flags:     setupFlags,
+		Nargs:     1,
+		UsageLine: "splore",
+	}
+
+	browse    = false
+	httpServe = http.Serve
+	mux       = &http.ServeMux{}
+	port      = 0
+)
+
+const indexHtml = `<!DOCTYPE html>
+<html>
+	<head>
+		<title>noms splore %s</title>
+		<style>
+		body { margin: 0; user-select: none; }
+		</style>
+	</head>
+	<body>
+		<div id="splore"></div>
+		<script src="/out.js"></script>
+	</body>
+</html>`
+
+func setupFlags() *flag.FlagSet {
+	flagSet := flag.NewFlagSet("splore", flag.ExitOnError)
+	flagSet.BoolVar(&browse, "b", false, "Immediately open a web browser.")
+	flagSet.IntVar(&port, "p", 0, "Server port. Defaults to a random port.")
+	verbose.RegisterVerboseFlags(flagSet)
+	return flagSet
+}
+
+type node struct {
+	nodeInfo
+	Children []nodeChild `json:"children"`
+}
+
+type nodeInfo struct {
+	HasChildren bool   `json:"hasChildren"`
+	ID          string `json:"id"`
+	Name        string `json:"name"`
+}
+
+type nodeChild struct {
+	Key   nodeInfo `json:"key"`
+	Label string   `json:"label"`
+	Value nodeInfo `json:"value"`
+}
+
+func run(args []string) int {
+	var sp spec.Spec
+	var getValue func() types.Value
+
+	cfg := config.NewResolver()
+	if pathSp, err := spec.ForPath(cfg.ResolvePathSpec(args[0])); err == nil {
+		sp = pathSp
+		getValue = func() types.Value { return sp.GetValue() }
+	} else if dbSp, err := spec.ForDatabase(cfg.ResolveDbSpec(args[0])); err == nil {
+		sp = dbSp
+		getValue = func() types.Value { return sp.GetDatabase().Datasets() }
+	} else {
+		d.CheckError(fmt.Errorf("Not a path or database: %s", args[0]))
+	}
+
+	defer sp.Close()
+
+	req := func(w http.ResponseWriter, contentType string) {
+		sp.GetDatabase().Rebase()
+		w.Header().Add("Content-Type", contentType)
+		w.Header().Add("Cache-Control", "max-age=0,no-cache")
+	}
+
+	mux.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
+		req(w, "text/html")
+		fmt.Fprintf(w, indexHtml, sp.String())
+	})
+
+	mux.HandleFunc("/out.js", func(w http.ResponseWriter, r *http.Request) {
+		req(w, "application/javascript")
+		// To develop JS, uncomment this line and run `yarn start`:
+		//http.ServeFile(w, r, "splore/out.js")
+		// To build noms-splore. uncomment this line and run `yarn buildgo`:
+		fmt.Fprint(w, outJs)
+	})
+
+	mux.HandleFunc("/getNode", func(w http.ResponseWriter, r *http.Request) {
+		req(w, "application/json")
+		r.ParseForm()
+		id := r.Form.Get("id")
+
+		var v types.Value
+		switch {
+		case id == "":
+			v = getValue()
+		case id[0] == '#':
+			abspath, err := spec.NewAbsolutePath(id)
```

Comment:
We might want to add a `spec.FigureOutPath` or something, but that would also need to give a clue into what it figured out (since we handle value vs absolute differently here, obviously).